domain/application/charm/store/store.go

```diff
@@ -57,16 +77,85 @@ func (s *CharmStore) Store(ctx context.Context, path string, size int64, hash st
 	// Generate a unique path for the file.
 	unique, err := uuid.NewUUID()
 	if err != nil {
-		return "", "", errors.Errorf("cannot generate unique path")
+		return StoreResult{}, errors.Errorf("cannot generate unique path")
 	}
 	uniqueName := s.encoder.EncodeToString(unique[:])
 
 	// Store the file in the object store.
+	objectStore, err := s.objectStoreGetter.GetObjectStore(ctx)
+	if err != nil {
+		return StoreResult{}, errors.Errorf("getting object store: %w", err)
+	}
+
 	uuid, err := objectStore.PutAndCheckHash(ctx, uniqueName, file, size, hash)
 	if err != nil {
-		return "", "", errors.Errorf("putting charm: %w", err)
+		return StoreResult{}, errors.Errorf("putting charm: %w", err)
+	}
+	return StoreResult{
+		ArchivePath:     file.Name(),
+		UniqueName:      uniqueName,
+		ObjectStoreUUID: uuid,
+	}, nil
+}
+
+// StoreFromReader stores the charm from the provided reader into the object
+// store. The caller is expected to remove the temporary file after the call.
+// IThis does not check the integrity of the charm hash.
+func (s *CharmStore) StoreFromReader(ctx context.Context, reader io.Reader, hashPrefix string) (StoreResult, Digest, error) {
+	file, err := os.CreateTemp("", "charm-")
+	if err != nil {
+		return StoreResult{}, Digest{}, errors.Errorf("creating temporary file: %w", err)
+	}
```

Comment:
Discussed offline.

We prefer storing the charm on disk, rather than memory to reduce memory pressure here.

However, we should `defer os.Remove(file.Name())` after we close the file, and stop leaking the file name.

This will take some changes to the charm service. See comment there.